flytekit/common/types/containers.py

```diff
@@ -149,3 +149,162 @@ def verbose_string(self):
             len(self.collection.literals),
             ",\n\t\t".join("\n\t\t".join(v.verbose_string().splitlines()) for v in self.collection.literals),
         )
+
+
+class MapType(_base_sdk_types.FlyteSdkType):
+    pass
+
+
+class TypedMapType(MapType):
+    @property
+    def sub_type(cls):
+        """
+        :rtype: flytekit.common.types.base_sdk_types.FlyteSdkType
+        """
+        return cls._sub_type
+
+    def __eq__(cls, other):
+        return hasattr(other, "sub_type") and cls.sub_type == other.sub_type
+
+    def __hash__(cls):
+        # Python 3 checks complain if hash isn't implemented at the same time as equals
+        return super(TypedMapType, cls).__hash__()
+
+
+def Map(sdk_type):
+    """
+    :param flytekit.common.types.base_sdk_types.FlyteSdkType sdk_type:
+    :rtype: flytekit.common.types.base_sdk_types.FlyteSdkType
+    """
+
+    class TMap(TypedMapImpl):
+        _sub_type = sdk_type
+
+    # TODO: Figure out generics and type-hinting
+    return TMap
+
+
+class MapImpl(_base_sdk_types.FlyteSdkValue, metaclass=MapType):
+    def __len__(self):
+        return len(self.map.literals)
+
+
+class TypedMapImpl(MapImpl, metaclass=TypedMapType):
+    @classmethod
+    def _ensure_key_is_str(cls, dictionary):
+        for k in dictionary:
+            if type(k) != str:
+                raise _user_exceptions.FlyteTypeException(
+                    type(k), str, k,
+                )
+
+    @classmethod
+    def from_string(cls, string_value):
+        """
+        Load the map from a JSON formatted string.
+        :param Text string_value:
+        :rtype: MapImpl<Text, T>
+        :raises: flytekit.common.exceptions.user.FlyteTypeException
+        """
+        try:
+            dictionary = _json.loads(string_value)
+        except ValueError:
+            raise _user_exceptions.FlyteTypeException(
+                _six.text_type, cls, additional_msg=f"String not parseable to json {string_value}",
+            )
+
+        if type(dictionary) != dict:
+            raise _user_exceptions.FlyteTypeException(
+                _six.text_type, cls, additional_msg=f"String is not a dictionary {string_value}",
+            )
+
+        cls._ensure_key_is_str(dictionary)
+
+        # Instead of recursively calling from_string(), we're changing to from_python_std() instead because json
+        # loading naturally interprets all layers, not just the outer layer.
+        return cls({k: cls.sub_type.from_python_std(v) for k, v in dictionary.items()})
+
+    @classmethod
+    def is_castable_from(cls, other):
+        """
+        :param flytekit.common.types.base_literal_types.FlyteSdkType other:
+        :rtype: bool
+        """
+        return type(cls) == type(other) and cls.sub_type.is_castable_from(other.sub_type)
```

Comment:
I'm not entirely sure this is the right way to implement. I checked the code of `list` but it didn't pass any unit test I tried.